Animation Templates Refactored [LLM assisted]

mentos987 opened the pull request:
Reworks how animation timers are handled. Done because I consider the old system hard to understand and inefficient in its time keeping.

Major changes:

1. Migrates animation time management to a wall-clock anchored system instead of each animation having a separate ticking timer. This means less risk of sync issues between animations since they now all use the same timer.
2. Renaming, commenting and moving logic around to organise animation time functionality.
3. Optimises 2 heavy functions. (animated<T>::advance_to_current_frame() and halo_impl::effect::update())

I tested every function I altered and mostly there is little change, except for two of them that got major gains.

Heavy test of **advance_to_current_frame()**. Done on max zoom out in an area with hundreds of animations going:
```
---- before:
[PerfTimer]: 132.028ms over 6299714 calls (avg 2.09577e-05ms)
[PerfTimer]: 134.676ms over 6708463 calls (avg 2.00756e-05ms)
[PerfTimer]: 137.335ms over 6972493 calls (avg 1.96966e-05ms)

--- after:
[PerfTimer]: 15.9779ms over 203170 calls (avg 7.86429e-05ms)
[PerfTimer]: 14.1693ms over 201247 calls (avg 7.04077e-05ms)
[PerfTimer]: 13.5699ms over 201596 calls (avg 6.73124e-05ms)
```

Light test of **advance_to_current_frame()**. Done with max zoom in on some grass tiles with 0 animations going:
```
---- before:
[PerfTimer]: 0.284861ms over 16640 calls (avg 1.71191e-05ms)
[PerfTimer]: 0.276722ms over 16380 calls (avg 1.68939e-05ms)
[PerfTimer]: 0.264466ms over 16484 calls (avg 1.60438e-05ms)

---- after:
none, we no longer process grass tiles as animations.
```

 Heavy test of **halo_impl::effect::update()**. Done on max zoom out in an area with hundreds of animations going:
```
---- before:
[PerfTimer]: 507.882ms over 1098306 calls (avg 0.000462423ms)
[PerfTimer]: 467.661ms over 1102004 calls (avg 0.000424373ms)
[PerfTimer]: 459.058ms over 1109400 calls (avg 0.00041379ms)

---- after:
[PerfTimer]: 338.51ms over 1113900 calls (avg 0.000303896ms)
[PerfTimer]: 339.352ms over 1113900 calls (avg 0.000304652ms)
[PerfTimer]: 339.453ms over 1113900 calls (avg 0.000304743ms)
```

mentos987 committed:
overlay animation added

This enables animations for normal overlays if an input path string is formatted like other animations ([14~20], or a list, ..). The changes mostly deal with parsing and reformatting strings.

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
New IPF: OFFSET

Adds a new image path function "OFFSET", that offsets an image by an about of pixels, used like: image=data/core/images/items/armor.png~SCALE(50, 50)~OFFSET(-8,18)

mentos987 committed:
Merge branch 'IPF-add-pixel-offset-input'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

Pentarctagon commented:
The `advance_to_current_frame()` results seem odd. It looks like it got worse? Averages are ~2 before and ~7 after.

mentos987 commented:
> The `advance_to_current_frame()` results seem odd. It looks like it got worse? Averages are ~2 before and ~7 after.

No, it got way better. The reason why the average call is slower is because we no longer get a ton of trash calls. We only call when there is an animation in need to progressing now. Call amount went from 6299714 to 203170 for the same test. @Pentarctagon 

It is somewhat misleading since the main optimisation was not to call it at all =P

Vultraz commented on src/animated.tpp:
```diff
@@ -16,244 +16,224 @@
-
-	// Update the index key to the appropriate frame for the current animation time.
-	auto iter = frames_.begin() + current_frame_key_;
-	while(std::next(iter) != frames_.end()) {
```
Where did this code go? We spent a ton of time fixing the perf issue related to it...

Vultraz commented:
The performance gains are great, and I really appreciate that you're chasing down inefficiencies in the render pipeline, but this PR is impossible to review as-is. It's a single commit wherein a *ton* of things have been moved without any explanation as to what is old (and simply renamed) and what's new, how it's been optimized, and how we can evaluate the changes to make sure they aren't causing any bugs. Our style guide is broken in numerous places, doc comments have been added to many functions that don't need them, class members have been removed or reordered without explanation...

Optimization should *not* be lumped in the same commit with cleanup or refactoring, and we need a clear explanation as to where the performance gains are coming from. You mentioned each animated object no longer keep its own clock - that's good! But I can't see how that manifests in the code you've given us. Given that AI was involved in producing it, we can't merge unless we can untangle what exactly Copilot or Claude or whatever tool is proposing here...

I'm honestly also concerned as to how much manual review is happening with these LLM-assisted changes. The fact that you're profiling and can give us hard performance numbers is excellent, as is the fact that you verify that everything renders correctly. But the formatting issues alone -- consistent ones at that -- make me think there hasn't been a manual review pass. A smoketest alone is not enough.

At the bare minimum, this PR needs to be multiple commits. An example (but by no means exhaustive) list:
- Renaming existing functions and class member for clarity without any other implementation changes
- Adding doc comments *where necessary* (not to every single function and class member!)
- Implementing the global clock system
- Optimizing advance_to_current_frame et al by reducing unnecessary calls
- Optimizing the halo effect

mentos987 commented:
I can separate out all the comments and the halo optimising into separate commits if you want. But following three points are very entangled (hard top split up): 
1.     Renaming existing functions and class member for clarity without any other implementation changes
2.     Implementing the global clock system
3.     Optimizing advance_to_current_frame et al by reducing unnecessary calls

"Copilot or Claude or whatever tool is proposing here" -- This is all my work, for better or worse the changes are not suggested by LLMs. I ask it to do rewrites functions by function according to how I want it changed.

"Our style guide is broken in numerous places" -- I'll do a pass of checking this.

"doc comments have been added to many functions that don't need them" -- How do I know what function need them? Seems arbitrary so I added to all. The old code took me 5+ hours to understand so it seemed woefully under explained.

"class members have been removed or reordered without explanation" -- Yes, in the template file I changed a lot. Because the old code was such a pain for me to get into so I thought to spare future users by doing a complete reorganisation into something easier to understand. I do get that the dif for that file is impossible to go through, you pretty much have to look at the file and debug in game. Which I do.

" how we can evaluate the changes to make sure they aren't causing any bugs" -- The changes are so extensive that in-game bug-testing is perhaps the simplest way to test. I do it by timing efficiency and by hunting down any visual artefacts in game. Since I know what animations use what functions I can test all the functions in game fairly easily.

"clear explanation as to where the performance gains are coming from" -- From two places:

1. Halo optimising comes from storing the pointer towards images in cache rather than having to recreate them all the time. A simple change.
2. The more extensive optimisation comes from reworking the animation system to no longer need individual timers for every image object. Before it did tick updates for all frames in update_last_draw_time() all the time. I removed that entirely. The  functionality it offered is now replaced by update_animation_timers(), which only needs to update 2 timers rather than 10000.

"formatting issues alone -- consistent ones at that -- make me think there hasn't been a manual review pass" -- I left it as I liked it for now. I have yet to go over it with CONTRIBUTING.md in mind. I'll do so.

I'll work towards your wishes as well as I can, thank you for the feedback @Vultraz

mentos987 commented:
My next steps will be: 

1. Test for more bugs.
2. Go over it with CONTRIBUTING.md
3. Look over all the comments again.
4. Split up in separate commits.

Vultraz commented:
Doc comments (and comments in general) are helpful when they convey information not otherwise conveyed elsewhere. They're a maintenance burden in and of themselves, so it's good to be judicious with comments. For instance, there's `get_first_frame`, which which the new comment is "Gets the first frame's value." That doesn't offer much beyond what the function name already tells you. (That's not to say such comments are always bad per-se. Sometimes you want to always have *some* sort of documentation, and I myself have done the same). But labeling the constructor as the constructor isn't necessary, nor is commenting on each member of the initializer list when you've already added comments to their declarations in the header.

I will say, though, most of the doc additions are good here. I would suggest using single-line format for short comments, `/** Like so. */`, to  reduce the amount of space they take up. (Obviously, multi-line comments are fine formatted as-is).

The most numerous style violations are the lines that look like this:
```cpp
if (frames_.empty()) {return starting_frame_time_;}
```
It should normally be:
```cpp
if(frames_.empty()) {
	return starting_frame_time_;
}
```
(Note we also don't put a space between `if` and its condition).

Wedge009 commented:
> Doc comments (and comments in general) are helpful when they convey information not otherwise conveyed elsewhere. They're a maintenance burden in and of themselves, so it's good to be judicious with comments. For instance, there's `get_first_frame`, which which the new comment is "Gets the first frame's value." That doesn't offer much beyond what the function name already tells you.

Agreed. I'm of the opinion that code references should be self-documenting and `get_first_frame` is an example of that. As a general rule, I follow/recommend the principle that comments explain the 'why' rather than the 'what/how'.

mentos987 committed:
Animation Templates Refactored

Reworks how animations are handled.

1. Migrates animation time management to a wall-clock anchored system instead of each animation having a separate ticking timer.

2. Renaming, commenting and moving logic around to organize everything.

mentos987 committed:
Implements the new animation system

Alters all the files that uses the animation templates.

Mostly name changes but a bit of logic change too.

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

Pentarctagon commented:
Should the `Optimization for halo_impl::effect::update()` commit be dropped?

mentos987 commented:
It is reverted in this PR as is. Moved to another PR and merged a while ago.

Pentarctagon commented:
Right, that's what I'm confused about. Why is there still a commit for it here when it was split to another PR?

mentos987 committed:
Merge remote-tracking branch 'upstream/master'

mentos987 commented:
Did something wrong on rebase and now the commit history looks bad. I don't know how to fix that but otherwise it seems fine.

Pentarctagon commented:
Is it possible to just drop everything but the first two commits? No idea why there are so many merge commits, but it looks like it also pulled it commits from other branches too (ie: the OFFSET IPF commit).

mentos987 commented:
Replaced with https://github.com/wesnoth/wesnoth/pull/11369 . Closing this PR.
